.github/workflows/tests.yml

```diff
@@ -125,25 +195,76 @@ jobs:
           node-version: lts/*
           cache: 'yarn'
 
+      - name: Restore cypress runner Cache
+        uses: actions/cache@v2
+        id: restore-cypress-cache
+        with:
+          path: ~/.cache/Cypress
+          key: ${{ runner.os }}-cypress-${{ hashFiles('yarn.lock') }}
+
+      - name: Restore node_modules cache
+        uses: actions/cache@v2
+        id: restore-cache
+        with:
+          path: |
+            ./node_modules
+            **/node_modules
+          key: ${{ runner.os }}-nodemodules-${{ needs.setup.outputs.commit }}
+
+      - name: Restore ui/dist cache
+        uses: actions/cache@v2
+        id: restore-ui-cache
+        with:
+          path: ./packages/ui/dist
+          key: ${{ runner.os }}-ui-${{ needs.setup.outputs.commit }}
+
+      - name: Restore ${{ matrix.package }}/dist cache
+        id: restore-package-cache
+        uses: actions/cache@v2
+        with:
+          path: ./packages/${{ matrix.package }}/dist
+          key: ${{ runner.os }}-${{ matrix.package }}-${{ needs.setup.outputs.commit }}
+
       - name: Install packages
-        run: yarn --frozen-lockfile
+        if: steps.restore-cache.outputs.cache-hit != 'true' || steps.restore-cypress-cache.outputs.cache-hit != 'true'
+        run: yarn --frozen-lockfile --prefer-offline
+
+      - name: Build @aws-amplify/ui package
+        if: steps.restore-ui-cache.outputs.cache-hit != 'true'
+        run: yarn ui build
+
+      - name: Build ${{ matrix.package }} package
+        if: steps.restore-package-cache.outputs.cache-hit != 'true'
+        run: yarn ${{ matrix.package }} build
 
       - name: Add Amplify CLI
         run: yarn global add @aws-amplify/cli
 
-      - name: Pull down AWS environments
+      - name: Get CLI versions
+        id: cli-version
+        run: echo "::set-output name=version::$(amplify --version)"
+
+      - name: Create or restore environments cache
+        id: environments-cache
+        uses: actions/cache@v2
+        with:
+          path: environments/**/aws-exports.js
+          key: ${{ runner.os }}-environments-${{ steps.cli-version.outputs.version }}-${{ hashFiles('environments/**/amplify/**') }}
```

Comment:
We reuse environments unless 1) CLI version and 2) `environments/**` contents change